File: 2025-maximum-number-of-ways-to-partition-an-array/2025-maximum-number-of-ways-to-partition-an-array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var waysToPartition = function(nums, k) {
    
    let leftCounter = {}
    let rightCounter = {}
    
    function getValue(obj, val) {
        if(obj[val] !== undefined) return obj[val]
        return 0
    }
    
    function addOne(obj, val){
        obj[val] = obj[val] || 0
        obj[val]++ 
        
    }
    
    function minusOne(obj, val){
        obj[val]-- 

        
    }
    let prefix = [...nums]
    
    for(let i  = 1; i < nums.length; i++){
        prefix[i] += prefix[i-1]
    }
    
    for(let i = 1; i < nums.length; i++){
        let diff = prefix[i-1] - (prefix[prefix.length -1] - prefix[i-1])
        addOne(rightCounter, diff)
    }
    
    
    let res = getValue(rightCounter, 0)
    
    
    for(let i = 0; i<nums.length; i++){
        let diff = k - nums[i]
        
        res = Math.max(getValue(rightCounter, -diff) + getValue(leftCounter, diff), res)
        let leftDiff = prefix[i] - (prefix[prefix.length-1] - prefix[i] )
        addOne(leftCounter, leftDiff)
        minusOne(rightCounter, leftDiff)

    }
    return res
    
};

File: 2025-maximum-number-of-ways-to-partition-an-array/2025-maximum-number-of-ways-to-partition-an-array.py
class Solution:
    def waysToPartition(self, nums: List[int], k: int) -> int:
        
        prefix = list(accumulate(nums))
        right = Counter()
        left = Counter()
        
        
        
        for i in range(1, len(nums)):
            diff = prefix[i-1] - (prefix[-1] - prefix[i-1])
            right[diff] += 1
            
        ans = right[0]
        
        for i in range(0, len(nums)) :
            d = k - nums[i]
            ans = max(right[-d] + left[d], ans)
            currDiff = prefix[i] - (prefix[-1] - prefix[i])

            right[currDiff] -= 1
            left[currDiff]  += 1
        return ans    